### PR TITLE
Pcap read directories recursively v6

### PR DESCRIPTION
**Pull request by jamesdutrisac**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/2363

Describe changes:

1. Added ability to recursively read pcap directories
2. src/suricata.c: addition of new command line parameter --pcap-file-recursive
3. src/source-pcap-file.c: parsing of the command line argument
4. src/source-pcap-file-directory-helper.h: two thread vars tracking directory depth and should recurse
5. src/util-error.c / src/util-error.h: Added new warning code "SC_WARN_PATH_READ_ERROR"
6.  src/util-path.c / src/util-path.h added 4 functions
    a) SCIsRegularFile to provide OS independent file info.
    b) SCIsRegularDirectory to provide OS independent directory info.
    c) SCRealPath is an OS independent wrapper for realpath.
    d) PathJoin to manage path resolution logic.

This updates version 5 (#5198) by moving the changes to util-path to their own commit.  As well, minor bug fixes and code updates as per the code review.

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):

#suricata-verify-pr:
#suricata-verify-repo:
#suricata-verify-branch:
#suricata-update-pr:
#suricata-update-repo:
#suricata-update-branch:
#libhtp-pr:
#libhtp-repo:
#libhtp-branch:

### CODE REVIEW

File: src/source-pcap-file.c [hunk 1]
```diff
@@ -292,11 +292,24 @@ TmEcode ReceivePcapFileThreadInit(ThreadVars *tv, const void *initdata, void **d
             CleanupPcapFileThreadVars(ptv);
             SCReturnInt(TM_ECODE_OK);
         }
+        pv->cur_dir_depth = 0;
+
+        int should_recurse = 0;
```
Comment (jlucovsky) [inline src/source-pcap-file.c:297]: nit: no need to initialize this variable -- `ConfGetBool` will do that.

File: src/source-pcap-file.c [hunk 2]
```diff
@@ -292,11 +292,24 @@ TmEcode ReceivePcapFileThreadInit(ThreadVars *tv, const void *initdata, void **d
...
+        }
 
         int should_loop = 0;
         pv->should_loop = false;
         if (ConfGetBool("pcap-file.continuous", &should_loop) == 1) {
-            pv->should_loop = should_loop == 1;
+            pv->should_loop = (should_loop == 1);
+        }
+
+        if (pv->should_recurse == 1 && pv->should_loop == 1) {
```
Comment (jlucovsky) [inline src/source-pcap-file.c:309]: Both `should_` vars are booleans

File: src/util-path.c [hunk 3]
```diff
@@ -66,6 +66,40 @@ int PathIsRelative(const char *path)
...
+#ifdef OS_WIN32
+#define DIRECTORY_SEPARATOR '\\'
+#else
+#define DIRECTORY_SEPARATOR '/'
+#endif
+    uint16_t max_path_len = (buf_len > PATH_MAX) ? PATH_MAX : buf_len;
+    int bytes_written = snprintf(out_buf, max_path_len, "%s%c%s", dir, DIRECTORY_SEPARATOR, fname);
+    if (bytes_written <= 0) {
+        SCLogError(SC_ERR_SPRINTF, "Could not join filename to path");
+        SCReturnInt(TM_ECODE_FAILED);
```
Comment (jlucovsky) [inline src/util-path.c:90]: Please add `SCEnter()`  to this function.

File: src/util-path.c [hunk 4]
```diff
@@ -66,6 +66,40 @@ int PathIsRelative(const char *path)
...
+ * \retval TM_ECODE_FAILED on failure
+ */
+TmEcode PathJoin (char *out_buf, uint16_t buf_len, const char *const dir, const char *const fname)
+{
+#ifdef OS_WIN32
+#define DIRECTORY_SEPARATOR '\\'
+#else
+#define DIRECTORY_SEPARATOR '/'
+#endif
+    uint16_t max_path_len = (buf_len > PATH_MAX) ? PATH_MAX : buf_len;
```
Comment (jlucovsky) [inline src/util-path.c:86]: suggest using `MAX` macro for readability

File: src/util-path.c (hunk 3 above)
Comment (jamesdutrisac) [reply src/util-path.c:90]: done.

**Comment by jamesdutrisac** [comment]
Closing for version 7: https://github.com/OISF/suricata/pull/5238

**Pull request closed without merging**